File: 2050-CountGoodNumbers/2050-CountGoodNumbers.cpp
// Last updated: 04/04/2026, 23:30:46
class Solution {
public:
    const long long MOD=1000000007;
    long long binExp(long long base,long long pow) {
        if(pow==0) return 1;
        long long ans = binExp(base,pow/2);
        ans = (ans*ans)%MOD;
        if(pow%2!=0) ans = (ans*base)%MOD;
        return ans;
    }
    int countGoodNumbers(long long n) {
        long long even = n/2+(n%2);
        long long odd = n/2;
        return (int)((binExp(5,even)*binExp(4,odd))%MOD);
    }
};

File: 2050-CountGoodNumbers/2050-CountGoodNumbers.java
// Last updated: 04/04/2026, 23:30:49
class Solution {
    public final long MOD = 1000000007;
    public long binExp(long base,long pow) {
        long ans = 1;
        while(pow >= 1) {
            if(pow%2==0) {
                base = (base * base) % MOD;
                pow = pow>>1; // same as pow = pow/2
            } else {
                ans = (ans * base) % MOD;
                pow--;
            }
        }
        return ans;
    }                                               
    public int countGoodNumbers(long n) {
        return (int)((binExp(5,(n+1)/2)%MOD*binExp(4,n/2)%MOD)%MOD);
    }
}